### ISSUE TITLE
Werid bounding boxes in NZTM

### ISSUE THREAD

**Issue by blacha** [developmentseed/morecantile#26]
When using NZTM TileMatrixSet the numbers appear to be off.

```
import morecantile

nztm = morecantile.tms.get('NZTM2000')
print(nztm.xy_bounds(morecantile.Tile(1, 2, 0)))
```

returns
```
CoordsBbox(xmin=12293760.0, ymin=-7881280.0, xmax=14587520.0, ymax=-5587520.0)
```

When viewing this tile inside of QGIS the top left point is ` { x: 1293760, y: 5412480 }`

I believe this is caused by https://github.com/developmentseed/morecantile/blob/master/morecantile/models.py#L305 as NZTM is defined as `[y, x]` rather than `[x, y]`

**Comment by blacha** [comment]
@vincentsarago I don't know of anywhere in the spec that determines the order,  

Inside of LINZ's basemap project we have currently just hard coded the XY order which is far from ideal :( 

https://github.com/linz/basemaps/blob/master/packages/geo/src/xy.order.ts#L24

Here are the definitions provided inside the TileMatrixSet

CRS: https://www.opengis.net/def/crs/EPSG/0/2193
Links to: https://www.opengis.net/def/cs/EPSG/0/4500 which defines [y, x]

**Comment by vincentsarago** [comment]
Alright I think I've got a start of a solution. 

I'm going to propose that we add `EPSGTreatsAsLatLong` and `EPSGTreatsAsNorthingEasting` to rasterio CRS class to be able to do like in GDAL

```
const bool bInvertAxis =
    oTargetSRS.EPSGTreatsAsLatLong() != FALSE ||
    oTargetSRS.EPSGTreatsAsNorthingEasting() != FALSE;
... 
const double dfOriX = bInvertAxis ? tmList[0].mTopLeftY : tmList[0].mTopLeftX;
const double dfOriY = bInvertAxis ? tmList[0].mTopLeftX : tmList[0].mTopLeftY;
```

https://github.com/OSGeo/gdal/blob/137640fc6f78a2fd91cff0d71163ad0b305c5104/gdal/frmts/gtiff/cogdriver.cpp#L241-L263

**Comment by sgillies** [comment]
Maybe pyproj can provide a workaround in the meanwhile? You can use WKT to convert rasterio CRS to pyproj CRS.

**Comment by vincentsarago** [comment]
Thanks @sgillies for the tips.
I also though about using pyproj instead of rasterio (but you know I'm a fan) too but I don't want to add pyproj in requirements because there is a huge shift in pyproj between versions and the latest one only support gdal>=3 and I don't want to risk having to rely on both gdal3 and gdal2 lib ( like rasterio wheels comes with gdal2 while pyproj might come with gdal3)

**Comment by palmerj** [comment]
> I'm going to propose that we add EPSGTreatsAsLatLong and EPSGTreatsAsNorthingEasting to rasterio CRS class to be able to do like in GDAL

Could just use GDAL python binding ;) However, with both approaches I think we need to ensure that PROJ >=6 is used. Before that the axis order is not defined in the SRS database.  Using pyproj >=2.0 I think would be ok as it internally uses PROJ 6.X which contains the axis order metadata.

**Comment by vincentsarago** [comment]
🎉 glad to see morecantile getting some interest 😄 

> Could just use GDAL python bindings ;)

I will take this as a joke ;-) 

looking at https://gdal.org/development/rfc/rfc20_srs_axes.html?highlight=epsgtreatsaslatlong#new-methods it seems the approach using  https://github.com/mapbox/rasterio/pull/1943 should work.


```python
from rasterio.crs import CRS, epsg_treats_as_northingeasting, epsg_treats_as_latlong 

def is_inverted(crs): 
    return not epsg_treats_as_latlong(crs) or not epsg_treats_as_latlong(crs)

crs = CRS.from_user_input("http://www.opengis.net/def/crs/EPSG/0/2193")                                                                         
is_inverted(crs)                                                                                                                                
> True
```

On my computer with PROJ version 5 ☝️

As I said ealier, I'd like to avoid adding more requirements and risking to have to rely on multiple PROJ/GDAL version so using rasterio first is my first choice

**Comment by palmerj** [comment]
OK looks good, and I double checked that no actual URL call is made with GDAL. I also checked further and GDAL < 3 does actually have pretty good code for dealing with axis order within importFromEPSGA (which used internally by CRS.from_user_input("http://www.opengis.net/def/crs/EPSG/0/XXX") ). Using GDAL 3 & PROJ6+ will be ok.

Note I think your is_inverted function is not correct:

```python
$ python
Python 3.7.7 (default, Mar 10 2020, 15:43:33)
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import gdal, osr
>>> gdal.VersionInfo()
'2040200'
>>> def invert_axis(crs):
...     return crs.EPSGTreatsAsLatLong() or crs.EPSGTreatsAsNorthingEasting()
...
>>> srs = osr.SpatialReference()
>>> srs.ImportFromEPSG(2193) # axis is stripped in GDAL <= 2.4
0
>>> invert_axis(srs)
0
>>> srs.ImportFromEPSGA(2193) # axis is added in GDAL <= 2.4
0
>>> invert_axis(srs)
1
```

So how about (un-tested):

```python
def is_inverted(crs): 
    return epsg_treats_as_latlong(crs) or epsg_treats_as_northingeasting(crs)
```

?

**Comment by vincentsarago** [comment]
Thanks @palmerj this is great! 

> I also checked further and GDAL < 3 does actually have pretty good code for dealing with axis order within importFromEPSGA (which used internally by 

Yeah the function to check if the CRS is inverted will only work for GDAL < 3 on the URI base CRS, but hopefully this is the format for the TMS specs. I'll make sure to reflect this in the Docs.

About the function, yeah it was wrong but is fixed in the PR https://github.com/developmentseed/morecantile/pull/27/files#diff-9ff327a7e57984e01623b07a8986dfffR28

Thanks a lot again

**Comment by palmerj** [comment]
> Yeah the function to check if the CRS is inverted will only work for GDAL < 3 on the URI base CRS, but hopefully this is the format for the TMS specs. I'll make sure to reflect this in the Docs.

Are you sure about this not working with GDAL 3? My testing with Python GDAL 3 and `srs.SetFromUserInput('http://www.opengis.net/def/crs/EPSG/0/XXXX')` works and provides the axis order.

**Comment by vincentsarago** [comment]
I'm saying using `CRS.from_epsg(2193)` in `.EPSGTreatsAsLatLong` or `.EPSGTreatsAsNorthingEasting()` will alway return False, but it was fixed in GDAL 3 as mentioned in GDAL docs https://gdal.org/api/ogrspatialref.html#classOGRSpatialReference_1ab1862f619ed2491a7ea9f7255fa3156e

tests in rasterio: https://github.com/mapbox/rasterio/pull/1943/files#diff-e108aea0f33c1fa74392017369c22c94R542-R546

So for GDAL 2, as long as the CRS is initiated from URI, the `is_axis_inverted ` will work.

**Comment by palmerj** [comment]
Yes correct. Or under GDAL 2 python (not available in rasterio) using the importFromEPSGA will also provide the axis order metadata.

**Comment by vincentsarago** [comment]
closed by #27 

will be included in morecantile==1.3.0
